### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by dlongley**
WIP, no need for further review yet

### CODE REVIEW

File: LICENSE.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+Bedrock Non-Commercial License v1.0
+===================================
+
+Copyright (c) 2011-2016 Digital Bazaar, Inc.
```
Comment (mattcollier) [inline LICENSE.md]: Bump date.

File: lib/config.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ */
+const config = require('bedrock').config;
```
Comment (mattcollier) [inline lib/config.js]: destructure

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ */
+require('bedrock-account');
```
Comment (mattcollier) [inline lib/index.js]: strict

File: lib/index.js [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+// module API
+const api = {};
+module.exports = api;
+
+const PERMISSIONS = bedrock.config.permission.permissions;
+const TOKEN_TYPES = ['password', 'nonce', 'challenge'];
+// base58 alphabet; used to improve human readability
+const ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
+
+const logger = bedrock.loggers.get('app').child('bedrock-authn-token');
```
Comment (mattcollier) [inline lib/index.js]: what are we doing about namespacing log stuff?  Adding `bedrock` everywhere seems redundant.  What else uses the bedrock logging system but bedrock modules?

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+ *
+ * @return a Promise that resolves to the token.
+ */
+api.get = brCallbackify(async ({actor, account, email, type}) => {
+  assert.optionalString(account, 'account');
+  assert.optionalString(account, 'email');
+  if(!(account || email)) {
+    throw new Error('Either "account" or "email" must be provided.');
+  }
+  validateTokenToken(type);
```
Comment (mattcollier) [inline lib/index.js]: terrible name.

File: lib/index.js [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+  if(account) {
+    query.id = database.hash(account);
+  } else {
+    query['account.email'] = email;
+  }
+  const record = await database.collections.account.findOne(
+    query, {_id: 0, [typeKey]: 1});
+  if(!record) {
+    throw new BedrockError('Authentication token not found.', 'NotFoundError', {
+      httpStatusCode: 400,
```
Comment (mattcollier) [inline lib/index.js]: 404 here and other places where `NotFoundError` is thrown?

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    id: database.hash(account)
+  }, {
+    $unset: {
+      [`meta.bedrock-authn-token.tokens.${type}`]: true
+    }
+  }, database.writeOptions);
+  if(result.result.n === 0) {
+    throw new BedrockError(
+      'Could not set account token. Account or token not found.',
+      'NotFoundError');
```
Comment (mattcollier) [inline lib/index.js]: httpStatusCode etc?

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+    }
+  }
+
+  if(type === 'challenge') {
+    // TODO: implement challenge check
+    // compare challenge results against challenge token; on success, write
+    // that the challenge was met to database
+    // possible challenges are small PoW like find factors to produce challenge
+    // number
+    throw new Error('Not implemented');
```
Comment (mattcollier) [inline lib/index.js]: period?

File: lib/index.js [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    };
+  }
+  return false;
+});
+
+// TODO: add API to send nonce token to account contact point (email/sms)
+api.notify = brCallbackify(async (
+  {actor, account, email, type, notification}) => {
+  console.log('notifying');
+  //throw new Error('Not implemented');
```
Comment (mattcollier) [inline lib/index.js]: enable throw?

File: lib/index.js [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+});
+
+function validateTokenToken(type) {
+  assert.string(type, 'type');
+  if(!TOKEN_TYPES.includes(type)) {
+    throw new Error('Token "type" must be one of: ' + TOKEN_TYPES.join(', '));
```
Comment (mattcollier) [inline lib/index.js]: This has to be a `TypeError` doesn't it!?  :)

File: package.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "bedrock-authn-token",
+  "version": "0.1.0-0",
+  "description": "Simple token-based authentication for Bedrock apps",
+  "main": "./lib",
+  "scripts": {},
```
Comment (mattcollier) [inline package.json]: remove?

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,11 @@
+*.log
+*.py[co]
+*.sw[nop]
+*~
+.project
+.settings
+TAGS
+bower_components
```
Comment (mattcollier) [inline .gitignore]: I'll save lehn the trouble

File: lib/index.js (hunk 4 above)
Comment (dlongley) [reply lib/index.js]: We decided to prune these (if necessary) in the internal logger code in the future.

**Comment by mattcollier** [comment]
@dlongley this OK for merge and 1.0.0 release?

**Comment by dlongley** [comment]
@mattcollier,

> this OK for merge and 1.0.0 release?

Nope. No one finished the notification parts. We need to make that part pluggable and test a mock notification thing that sets a var in a test. Then we can release 1.0 and add notification plugins (like email/sms) later.

**Comment by dlongley** [comment]
We meed to call `api.notify` here instead of a `console.log` of the nonce:

https://github.com/digitalbazaar/bedrock-authn-token/blob/implementation/lib/index.js#L74

And then have `api.notify` do some plugin thing like we do in other places.

**Pull request closed without merging**